AlgebraInLean/World0LeanIntroSOLUTIONS/Sheet2Solutions.lean

```diff
@@ -0,0 +1,283 @@
+  apply h2
+  intro h4
+  apply h1
+  intro h5
+  exact h4
+  done
+
+/-FM Section 1 Sheet 1-/
+example : ((Q → P) → P) → (Q → R) → (R → P) → P := by
+  intro h1 h2 h3
+  apply h1
+  intro h4
+  apply h3
+  apply h2
+  exact h4
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (((P → Q) → Q) → Q) → P → Q := by
+  intro h1 h2
+  apply h1
+  intro h3
+  apply h3
+  exact h2
+  done
+
+/-FM Section 1 Sheet 1-/
+example :
+    (((P → Q → Q) → (P → Q) → Q) → R) →
+      ((((P → P) → Q) → P → P → Q) → R) → (((P → P → Q) → (P → P) → Q) → R) → R := by
+  intro h1 h2 h3
+  apply h2
+  intro h4 h5 h6
+  apply h4
+  intro h7
+  exact h5
+  done
+
+/-The ↔ means "if and only if." If you are ever curious how to type a certain symbol in Lean, just
+hover over it for a few seconds. rw can be applied to ↔ hypotheses.-/
+
+/-FM Section 1 Sheet 5-/
+example : (P ↔ Q) → (Q ↔ P) := by
+  intro h
+  rw [h]
+  done
+
+/-FM Section 1 Sheet 5-/
+example : (P ↔ Q) → (Q ↔ R) → (P ↔ R) := by
+  intro h1 h2
+  rw [h2] at h1
+  exact h1
+  done
+
+/-Note that many of the exercises above could be solved in one line, using the tactic "tauto," just
+as some levels in NNG could be solved by "tauto."-/
+
+/-Lean also has booleans True and False.
+
+If the goal is "True", it can be solved with the tactic "trivial".
+
+For example:-/
+
+/-FM Section 1 Sheet 2-/
+example : True := by
+  trivial
+  done
+
+/-Let's do some exercises, adding "trivial" to our list of tactics.-/
+
+/-FM Section 1 Sheet 2-/
+example : True → True := by
+  intro h1
+  trivial
+  done
+
+/-FM Section 1 Sheet 2-/
+example : False → True := by
+  intro h1
+  trivial
+  done
+
+/-FM Section 1 Sheet 2-/
+example : False → False := by
+  intro h1
+  exact h1
+  done
+
+/-FM Section 1 Sheet 2-/
+example : (True → False) → False := by
+  intro h1
+  apply h1
+  trivial
+  done
+
+/-FM Section 1 Sheet 2-/
+example : True → False → True → False → True → False := by
+  intro h1 h2 h3 h4 h5
+  exact h2
+  done
+
+/-FM Section 1 Sheet 2-/
+example : P → (P → False) → False := by
+  intro h1 h2
+  apply h2 at h1
+  exact h1
+  done
+
+/-The tactic "exfalso" changes any goal after the ⊢ symbol to "False."
+
+For example:-/
+
+/-FM Section 1 Sheet 2-/
+example : False → P := by
+  intro h1
+  exfalso
+  exact h1
+  done
+
+/-Let's do some exercises.-/
+
+/-FM Section 1 Sheet 2-/
+example : (P → False) → P → Q := by
+  intro h1 h2
+  apply h1 at h2
+  exfalso
+  exact h2
+  done
+
+/-FM Section 1 Sheet 2-/
+example : (True → False) → P := by
+  intro h1
+  exfalso
+  apply h1
+  trivial
+  done
+
+/-To wrap up this part, note that the more advanced form of the "rw" tactic is "nth_rewrite". This
+example from the NNG shows it in use:-/
+
+/-NNG Advanced Addition World Level 3-/
+example (x y : Nat) : x + y = y → x = 0 := by
+  intro h1
+  nth_rewrite 2 [← zero_add y] at h1
+  apply add_right_cancel at h1
+  exact h1
+  done
+
+/-Note that tactics "exact" and "apply" both have variants, "exact?" and "apply?", which both go
```

Comment:
It'd be good to spell out the differences between the two better.
* `exact?` looks for theorems that *close* the goal.
* `apply?` looks for theorems that can be `apply`'d, but don't necessarily close the goal.
  * also would be nice to mention that this one is hit or miss. It won't work at all if you need to use `rw`.